### ISSUE TITLE
Path Completion on Common Prefixes

### ISSUE THREAD

**Issue by mitnk** [xonsh/xonsh#1397]
Say we have the following dirs in pwd:

```
$ ls -F
foo/    foo bar/
```

Then in bash:

```
$ cd f<TAB>
```

would get:

```
$ cd foo
```

While in xonsh, it remains with `cd f` because xonsh currently returns `{"'foo bar'", "foo"}`, so that readline treat them as different prefix (they are different).

I think we should make it same as bash.

Thoughts?

**Comment by PaulReiber** [comment]
My first thought is, I think your example is very good for demonstrating why filenames just plain shouldn't have spaces in them. Spaces already have an important job to do, separating commandline arguments from each other, and when you put them into filenames it introduces a bunch of special cases.  

Let's consider.  You type 'f', then tab.   It completes to "foo".  You type a SPACE character, then tab.

How could the completer possibly know if you meant...
 (a) I'm happy with foo on my commandline, and I'd like to complete '' (empty string) for the next argument (which is a totally valid thing to do), 
or
 (b) Although foo is a totally valid and existing directory, I want the completer to throw in whatever's needed ( a \ before the space?) and complete 'foo bar' for me.

Should an initial tab do (a) then a second tab do (b)?  Special cases.  Yuck.

I can't say I agree regarding making it the same as bash.  I'm for simplifying this instead.

What's in bash is pretty complex - the 'programmable completion' section alone, in the bash manual, is 80+ lines long, and a rough check shows around a dozen end-user-preference variables which affect its operation.  

While I do think xonsh needs powerful tab-based completion... I'm not at all sure that mimicking what's in bash is the right approach.

**Comment by mitnk** [comment]
IMO, at least we need to complete `f` to `foo`, what next up to the user. This is more natural for any shell I think (because they are with the same prefix, any shell doesn't complete like this?).

You're right, xonsh speak a different language, even `cd foo\ bar` don't work here (yet). The thing is I don't know whether this is a `will-support` or a `won't-support`. If latter, xonsh may need to put a word into the docs to tell users/devs the philosophy behind and probably a `won't-fix` list.

I think it's not about complex or simple, it's about what's the right thing. For me, 1) completes `f` to `foo` is the right thing, 2) support `foo\ bar` or not is choosable, up to xonsh own philosophy.

**Comment by mitnk** [comment]
Seems latest master sort of fixed (improved) this issue. `cd f<TAB>` will be `cd 'foo`, which fix my complains! I tried but didn't find the related PR fixing this.

Anyway, I will close this issue if no discussions.

**Comment by scopatz** [comment]
I think xonsh is doing the right thing here but not for any space related reason.  Rather foo and "foo bar" both start with the same sub-string (namely foo), so hitting tab will fill in the longest common substring.  

Reimagining this problem, if I have two dirs, `foobar` and `"foo bar"`, and then I run `cd f<TAB>`, I get `cd foo`, like I should.

Note that in the original problem, if "foo bar" didn't exist, the completion would have a trailing `/` and it doesn't.

**Comment by mitnk** [comment]
> if "foo bar" didn't exist, the completion would have a trailing / and it doesn't.

I got trailing `/` on my side, both readline and PTK.

**Comment by scopatz** [comment]
@mitnk can you maybe please make a little asciinema please? I am not sure what is going on exactly

**Comment by mitnk** [comment]
@scopatz the part about trailing `/`?

When there is only one directory `foo` and no other dirs starts with `foo`, you said (if i understand correctly) when typing `$ cd f<TAB>`:

> the completion would have a trailing / and it doesn't.

On my side `$ cd f<TAB>` will complete to `$ cd foo/`.

On this whole issue, I was suggesting that when there are two dirs: `foo` and `foo bar`. xonsh should completes `$ cd f<TAB>` to `$ cd foo` (completes up to the max common prefix), while xonsh currently is do nothing when I typing `$ cd f<TAB>` (with readline shell).

**Comment by scopatz** [comment]
> while xonsh currently is do nothing when I typing $ cd f<TAB> (with readline shell).

That is not the behaviour I see with the readline shell. I get the max common prefix.  You are on Mac, right?  Are you using gnureadline or libedit?

**Comment by mitnk** [comment]
Hmm.. I have the save behaviour on Mac and a Debian. I have gnureadline installed on Mac. Did you mimic on the case in first thread - two dirs: `foo` and `foo bar`?

**Comment by scopatz** [comment]
yes I mimiced that case.  what is your current xonfig? and can you please create an asciinema

**Comment by mitnk** [comment]
xonfig:
```
+------------------+--------------+
| xonsh            | 0.4.3.dev209 |
| Git SHA          | gd32819f     |
| Python           | 3.5.2        |
| PLY              | 3.7          |
| have readline    | True         |
| prompt toolkit   | 1.0.3        |
| shell type       | readline     |
| pygments         | 2.1.3        |
| on posix         | True         |
| on linux         | False        |
| on darwin        | True         |
| on windows       | False        |
| on cygwin        | False        |
| is superuser     | False        |
| default encoding | utf-8        |
+------------------+--------------+
```

here is [the asciinema](https://asciinema.org/a/1q5u46xugro8wljcrly6cm9h5)

**Comment by scopatz** [comment]
ok, thanks a ton!  Are you using libedit or gnureadline. in xonsh, what does the following print,

```
import readline
readline?
```

**Comment by mitnk** [comment]
gnureadline i guess:
```
$ readline?
Type:         module
String form:  <module 'gnureadline' from '/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/gnureadline.cpython-35m-darwin.so'>
File:         /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/gnureadline.cpython-35m-darwin.so
Docstring:    Importing this module enables command line editing using GNU readline.
<module 'gnureadline' from '/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/gnureadline.cpython-35m-darwin.so'>
```

**Comment by scopatz** [comment]
hmmm weird... what version of readline is it?

**Comment by mitnk** [comment]
6.3.3 latest. here is my [xonshrc](https://github.com/mitnk/dotfiles/blob/master/xonsh/xonshrc) if you want to check. But I cannot image it can be completed to `foo` because it literally not with the same prefixes as xonsh returns(one is `foo`, another `'foo`, as the asciinema shows).

**Comment by gforsyth** [comment]
Hmm, yeah, I can confirm this on both PTK and readline on my end -- it's only an issue when the dirs have spaces since the spaced directory is prepended with the `'`.  

That's annoying.

**Comment by gforsyth** [comment]
Hmm, yeah, I can confirm this on both PTK and readline on my end -- it's only an issue when the dirs have spaces since the spaced directory is prepended with the `'`.  

That's annoying.

**Comment by scopatz** [comment]
Hmm maybe this is working for me because autoquoting is not working at all!

**Comment by blahgeek** [comment]
I'm using PTK and seems pretty good to me:

<img width="137" alt="screen shot 2016-10-25 at 22 15 16" src="https://cloud.githubusercontent.com/assets/1308450/19689675/babe1f5c-9b00-11e6-9fcc-711147cb90c2.png">
<img width="122" alt="screen shot 2016-10-25 at 22 15 24" src="https://cloud.githubusercontent.com/assets/1308450/19689678/bc232c02-9b00-11e6-967e-fd99152f6e12.png">
<img width="135" alt="screen shot 2016-10-25 at 22 15 30" src="https://cloud.githubusercontent.com/assets/1308450/19689681/bde966a0-9b00-11e6-82ce-0d35f6f8d4bd.png">

(So why are we need to support both readline and PTK anyway? I don't think there's any reason not to use PTK and there're many issues related to readline only.)

**Comment by gforsyth** [comment]
Hey @blahgeek -- some users prefer the `readline` shell and it's also faster, which is helpful for people on slower machines.

**Comment by gforsyth** [comment]
Fixed in #1897
